Add Reconciliation Metrics Functionality

vanou opened the pull request:
This commit
* adds two metrics
  - Count the number of reconciliation & failure of reconciliation
  - Track the number of objects in manifest

* modifies walkthrough example to mention this metrics functionality

* and deletes
   * reconcilerParams.groupVersionKind option
   * WithGroupVersionKind function

Added two metrics are registered to controller-runtime Prometheus registry

This PR is related to #65 .

rajansandeep commented:
/ok-to-test

vanou commented:
Thanks @atoato88 for reviewing!

I updated PR.

johnsonj commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,272 @@
+		ns = DefaultNamespace
+	}
+
+	if gvkt, ok := ot.trackedGVK[gvk]; ok {
```
I think you'll want to add a lock/mutex around trackedGVK. It is possible to have parallel reconciles if the user enables it in controller-runtime[1]
 
[1] https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller?tab=doc#Options

johnsonj commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -63,29 +69,44 @@ type DeclarativeObject interface {
+		return
+	}
+
+	metrics.Registry.MustRegister(reconcileCount, reconcileFailure, managedObjectsRecord)
```
Can you make Metrics tracking an optional (default off) feature then update the [example](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/tree/master/docs/addon/walkthrough#using-the-framework-in-the-controller) to turn it on?

This is the pattern we've been following (eg with [WithApplyPrune](https://pkg.go.dev/sigs.k8s.io/kubebuilder-declarative-pattern@v0.0.0-20200522144838-848d48e5b073/pkg/patterns/declarative?tab=doc#WithApplyPrune)). This allows us to keep a stable API/functionality for users while introducing new features.

I expect, not so far in the future, we have a 'Recommended Set of Options' and I would certainly expect this to be apart of that. For now, our recommended set is the ones in the walkthrough.

johnsonj commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -337,6 +362,43 @@ func (r *Reconciler) validateOptions() error {
+				}
+			}
+		}
+		// Are there any function to recognize K8s API versioning?
```
I'm not sure, @justinsb are you aware of anything "upstream" to resolve the api version here?

vanou replied:
Yes.

When
* `Scheme.ObjectKinds` returns multiple `GruopVersionKind` &
* `Scheme.PrioritizedVersionsForGroup` doesn't help determine which version to use,

we need a way to recognize K8s API versioning & choose appropriate (more matured) version.

johnsonj commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,272 @@
+
+package declarative
+
+import (
```
Could you add test coverage for this file?

johnsonj commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -63,29 +69,44 @@ type DeclarativeObject interface {
 var kubectl = kubectlcmd.New()
 
-func (r *Reconciler) Init(mgr manager.Manager, prototype DeclarativeObject, opts ...reconcilerOption) error {
+func (r *Reconciler) Init(mgr manager.Manager, prototype DeclarativeObject, opts ...reconcilerOption) (err error) {
```
super nit: can you drop the named return here and in setReconcileMetrics? it makes it a bit harder to read the function once it's more than few lines.

vanou replied:
OK. I'll drop named return in `*Reconciler.Init` here. However, in `setReconcileMetrics`, we cannot use short variable declaration [here](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/pull/98/files#diff-23c5c12ad7f72be6835d4f216e702363R370) because `gvks` will be redeclared & be in scope of 'if' block.

Should I use function scope variable `err`?

johnsonj replied:
In the remaining cases in setReconcileMetrics you should `return nil`, this makes it clear when reading the code that no error can be returned from this valid path.

vanou replied:
Thanks. I understand.

vanou commented:
Thanks @johnsonj for review! I'll work on your feedback.

johnsonj commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -337,6 +362,43 @@ func (r *Reconciler) validateOptions() error {
+					if gv.Version == check.Version {
+						r.rc = reconcileCounterFor(check)
+						r.rfc = reconcileFailureCounterFor(check)
+						return
```
is return the right  decision here? I'm having a hard time understanding what's going on in this block. Can you add a comment?

vanou replied:
In `setReconcileMetrics`, I would like to know GVK of `Reconciler.prototype` to set metrics' `group_version_kind` label value. However `prototype` argument of `Reconcler.Init` is usually address of empty CRD type and `prototype.GetObjectKind().GroupVersionKind()` doesn't return helpful information.

In this commit, I try to determine GVK from `Scheme`. However there is a chance controller manager's `Scheme` returns multiple version of GVK holding same CRD type. To determine which version to use, I use `PrioritizedVersionsForGroup`. If GVK `scheme.PrioritizedVersionsForGroup` returns matches one of GVKs returned by `scheme.ObjectKinds`, we use that GVK as `group_verison_kind` label value. If not, we use `gvks[0]`.

However, if setting [`reconcilerParams.groupVersionKind`]( https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/master/pkg/patterns/declarative/options.go#L42) becomes mandatory (`Reconciler.validateOptions` check it), there is no need to work on controller manager's `Scheme` to determine GVK.

johnsonj replied:
Ah! Thank you for the explanation @vanou - You've uncovered an issue here! WithGroupVersionKind[1] states 'This option is required'[1] but there is no validation around that, and no operators are specifying it! (checked this repo and kubernetes-sigs/cluster-addons).

For this code, you should use apiutil.GVKForObject(..)[2] and we should just delete the unused `reconcilerParams.groupVersionKind` / `WithGroupVersionKind` option. 

[1] https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/dfc7666ed8e6583e6a37fe39c08d9471aaa241d0/pkg/patterns/declarative/options.go#L95
[2] https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/ce639cbf6d4ce402273ad35ecd5f2996758b4f6b/pkg/patterns/declarative/reconciler.go#L367

vanou replied:
OK. Thanks!

vanou commented:
Hi @johnsonj .

I dealt with your feedback. Thank you! In this updated commit, I add one functionality to recognize whether K8s object is namespaced or cluster-scoped.

k8s-ci-robot says *pull-declarative-test* failed. This is because I use [envtest package](https://github.com/kubernetes-sigs/controller-runtime/tree/master/pkg/envtest) in sigs.k8s.io/controller-runtime. This package requires kube-apiserver, etcd & kubectl binary to run local kube-apiserver and against it we can run test.

justinsb commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -64,28 +72,69 @@ type DeclarativeObject interface {
+				}
+			}()
+		} else {
+			log.WithName("declarative_reconciler").Error(err, "failed to get corresponding RESTMapper from API server. try next time")
```
Nit: I'm not sure what "try next time" is implying ... does it mean "will retry" ?

vanou replied:
Yes. This means "will try". I'll modify it to "Reconciler will try to get RESTMapper next time."

justinsb commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,342 @@
+		Subsystem: Declarative,
+		Name:      ManagedObjectsRecord,
+		Help:      "Track the number of objects in manifest",
+	}, []string{"group_version_kind", "namespace", "name"})
```
My understanding is that we should generally avoid unbounded dimensions, because of the limits of prometheus scalability.  But this advice may be a few years out of date ... is it still true?

I don't know if we would classify these as unbounded, because in general they are going to be specified in the manifests.  One possible problem is if we started hashing configmaps (e.g. using kustomize), then the number of names would grow.

I want to see these metrics though - could we / should we start with just the group_version_kind dimension?

vanou replied:
@justinsb 
One approach to this is to introduce aging to pair of `group_version_kind`, `namespace` & `name`. If a triad hasn't been kubectl-applied but reconciliation has happened *n* times, that triad is removed from prometheus metrics. `n` can be set by any number when `WithReconcileMetrics` is called.
Is it reasonable?

justinsb commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,342 @@
+	}
+}
+
+type reconcileCounter struct {
```
One thing you could here is to have something like `reconcileMetrics` and expose a `ReconcileSuceeded` and `ReconcileFailed` method, and not expose the metric structure externally.  Then e.g. you could have `ReconcileFailed` take an error, if in future you wanted to handle some errors differently (a different metric) you could do that entirely in this metrics layer.

vanou replied:
I'll create new type `reconcileMetrics` in `metrics.go` file and use it. That type will implement similar method you suggest.

justinsb commented on pkg/patterns/declarative/options.go:
```diff
@@ -92,15 +92,6 @@ func WithObjectTransform(operations ...ObjectTransform) reconcilerOption {
-// This option is required.
-func WithGroupVersionKind(gvk schema.GroupVersionKind) reconcilerOption {
-	return func(p reconcilerParams) reconcilerParams {
-		p.groupVersionKind = &gvk
```
Nice cleanup - thanks!

justinsb commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -391,6 +448,17 @@ func (r *Reconciler) injectOwnerRef(ctx context.Context, instance DeclarativeObj
 }
 
+func (r *Reconciler) tryToSetCounter() error {
+	if mapping, err := r.mgr.GetRESTMapper().RESTMapping(r.gvk.GroupKind(), r.gvk.Version); err == nil {
```
Possible simplification:  rather than have the caller check `setCounter`, you could check in this function for the cached `rc` and `rfc` values, and return them.

Also I think you need a mutex guarding `setCounter` because I think Reconcile can be shared between goroutines.

If you're able to refactor `rc` and `rfc` into the same object, I would just store it in a single field e.g. `reconcileMetrics`.  Then I think we can change this method to be `getReconcileMetrics() (*reconcileMetrics, error)`

vanou replied:
I'll create type `reconcileMetrics` in `metrics.go` and use it. And I'll stop checking whether CRD is namespaced through `RESTMapper`, because I'll leave namespace as empty value if `reconcile.Request.Namaspace` is empty.

justinsb commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -64,28 +72,69 @@ type DeclarativeObject interface {
 	ctx := context.TODO()
 	log := log.Log
 
+	if r.CollectMetrics() {
```
Could you replace this with `defer r.collectMetrics(request, result, err)` and isolate the logic into a separate function

vanou replied:
OK. I'll do it.

justinsb commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -143,6 +192,14 @@ func (r *Reconciler) reconcileExists(ctx context.Context, name types.NamespacedN
 
+	if r.CollectMetrics() {
+		for _, obj := range objects.Items {
+			if err := globalObjectTracker.addIfNotPresent(obj); err != nil {
```
You could move this below the `ns` population logic, and then you can pass in the default namespace into `addIfNotPresent`

vanou replied:
Thanks. I overlooked that `ns` population logic. I'll do it.

justinsb commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,342 @@
+			ns = DefaultNamespace
+		}
+	} else {
+		ns = ClusterScoped
```
Does prometheus allow the empty value?  I suspect that would be more intuitive if so.

vanou replied:
Promtheus allows empty value. However, here, I need something to distinguish namespaced object and cluster-scoped object. 
I'll set metrics namespace label to empty value if namespace is not specified because it isn't reasonable to say empty namespace is `default` namespace as you point out in another comment.

justinsb commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,342 @@
+
+	if namespaced {
+		if len(ns) == 0 {
+			ns = DefaultNamespace
```
You need to take the default namespace as an arg, I think, because it won't always be `default`

vanou replied:
As your comment on `ns` population logic, you're right. I'll deal with it. Thanks.

justinsb commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,342 @@
+	return ot
+}
+
+type gvkTracker struct {
```
This is impressive, but a lot of work.  What do we think this metric will be used for?  Should we track a counter of the number of applies instead (which might be easier, and give similar results, depending on what we want it for)?

justinsb commented:
Thanks @vanou  - some good stuff.  I had some feedback and some questions, particularly around the cardinality and what we think the per-objects counts will be used for.  I'm wondering if we should just start with something simpler (e.g. a count of apply operations, per gvk)...

vanou commented:
Thanks @justinsb . I'll work on some of your recommendation.

About what metrics should we have, I think, tracking number of apply per gvk is not useful. This won't tell what objects is deployed by current channel. @johnsonj , do you have any comment for metrics functionality roadmap?

justinsb commented:
One compromise we could make is to put the applied metrics behind a "flag" - i.e. not always on.  So you (and others) can still have it, but we're not creating the unbounded metrics by default for other operators.  Would that be reasonable?

vanou commented:
It's reasonable for me. However, in my understanding, `WithReconcileMetrics` function in `options.go` seems enough. Could you tell me what should I do to put metrics behind flag?

vanou commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,342 @@
+func (rc reconcileCounter) reconcileWith(request reconcile.Request) {
+	if rc.namespaced {
+		if len(request.Namespace) == 0 {
+			request.Namespace = DefaultNamespace
```
You are right. I'll leave namespace as empty if `reconcile.Request` has empty namespace value. And because of it, I remove code checking whether CRD is namespaced through `RESTMapper`.

vanou commented:
Hello @justinsb . I modified PR following your recommendation.
However I added one feature which enables deleting metrics of removed k8s objects after user-specified times reconciliation happened. I think this feature solves unbounded metrics problem.

As I mentioned above, CI test failed because of [envtest package](https://github.com/kubernetes-sigs/controller-runtime/tree/master/pkg/envtest) used in unit test.

Your comment is very helpful. Thank you.

justinsb commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,600 @@
+type gvkTracker struct {
+	list     *items
+	src      source.Source
+	evthdler handler.EventHandler
```
Nit: eventHandler and predicate; shorter variable names are harder to read (and probably not _that_ much faster to type any more :-) , particularly with autocomplete)

justinsb commented on pkg/patterns/declarative/metrics.go:
```diff
@@ -0,0 +1,600 @@
+	return it
+}
+
+type bol map[string]*counter
```
Nit: I'm not sure what bol is short for, the full name would be clearer

vanou replied:
First I use `bol` as abbreviation of `bookOfLife`. However I'll use `record` instead of `bol`.

justinsb commented:
So as this is optional (thanks for pointing that out), I think it's reasonable to merge this in and we can iterate on it as needed :-)

Some style nits: I don't _think_ we normally abbreviate variable names (obvious exceptions are method receivers, range variables and often the word `obj` for object ! )

/approve
/lgtm

vanou commented on pkg/patterns/declarative/reconciler.go:
```diff
@@ -18,6 +18,7 @@ package declarative
 
 import (
 	"context"
+	stdlibErrs "errors"
```
I use `errors` to standard package & use `apierrors` to `k8s.io/apimachinery/pkg/api/errors`. All lowercase as `apiutil` for `sigs.k8s.io/controller-runtime/pkg/client/apiutil`.

vanou commented:
Thanks @justinsb . I modified names of variables.

> So as this is optional (thanks for pointing that out), I think it's reasonable to merge this in and we can iterate on it as needed :-)

*(Just to avoid miscommunication)*
Yes, this feature is optional. However I improved PR to address unbounded metrics problem, so I think it's bit fair to use this feature without concern. (Of course, there may be points we should improve.)

vanou commented:
Hi @justinsb . I rebased this PR. However, for my fault, this PR contains 4 commits now. Should I do any workaround?

justinsb commented:
Yeah, would you mind squashing the commits @vanou ?  We don't always have to - it can be nice to divide a big commit into logical steps - but in this case the commits seem mostly to be merging/rebasing, so squashing would be great!

vanou commented:
> Yeah, would you mind squashing the commits @vanou ? We don't always have to - it can be nice to divide a big commit into logical steps - but in this case the commits seem mostly to be merging/rebasing, so squashing would be great!

Thanks @justinsb . I tried to squash and because of need of resolving merge conflict, it resulted in 2 commits. And I think this PR is ready.

justinsb commented:
Thanks @vanou :tada: 

/approve
/lgtm
